### DISCUSSION TITLE
PowerShell 脚本启动

### DISCUSSION THREAD

**Discussion by loktindyi** [Melledy/Grasscutter#1284] [Show and tell]
每次都要各种点点点才能进入割草机，于是我写了一个脚本来快速进入割草机
现在这个脚本拥有以下功能：

- [x] 自我窗口隐藏（完全隐藏可能被安全软件拦截）
- [x] 自动开启系统代理（仅切换开关，代理地址需要提前手动填好）
- [x] 自动关闭系统代理（下面细说）
- [x] 请求打开原神（下面细说）
- [x] 打开割草机（启动服务端）
- [x] 打开mitmdump
- [x] mitmdump窗口隐藏（也可以不这样）
- [x] server窗口最小化（也可以不这样）

### 关于请求打开原神
我的电脑开启了UAC，即打开原神前会弹窗确定打开
如果取消，则原神打开失败，脚本则会执行关闭mitmdump和系统代理
这样做是为了方便在意外中(比如提前关闭了倒计时窗口)未能正确关闭代理时，可以以这种方式轻易地关闭代理
如果没有开启UAC，这个功能就不管用了

### 关于自动关闭系统代理
出现大门显示“开始游戏”字样时，登陆就已经成功，此时可以关闭代理了
自原神启动开始计时到登陆成功，这个时间与个人电脑配置相关
`$loginTimeout`指定从原神启动后多少秒内关闭代理，90秒一般够用
你可以依据你所花费的时间，再加上20秒留作意外情况的延时，来设置你的登录超时
在设置完成后，你可以选择将下面代码中第2行中的`Minimized`改为`Hidden`
这样就不会显示关闭代理的倒计时窗口了
> 其实，隐藏不隐藏都一样，感知不强

### 关于窗口隐藏
- 你可以将第2行中的`Minimized`改为`Hidden`，这样就不会显示关闭代理的倒计时窗口了（不太推荐）
`Minimized`将使脚本程序（倒计时窗口）以默认最小化运行，这已经足够隐蔽了，反正几十秒后就会关闭。而`Hidden`将完全隐藏窗口，这个行为可能被安全软件认定为恶意行为，我不推荐你因此而关闭安全软件设置。
- 你可以将第20行中的`Normal`改为`Minimized`，这样就可以将服务端窗口默认以最小化运行
- 你可以将第22行中的`Hidden`改为`Minimized`或者`Normal`，这样就可以...还是不改吧

### 代码
记得自己改前面的几项参数
```POWERSHELL
param($self = $true)
if ($self) { Start-Process PowerShell -ArgumentList ('-WindowStyle Minimized "{0}" $false' -f $MyInvocation.MyCommand.Definition); Exit }
# 登陆超时
$loginTimeout = 90
# 倒计时刷新间隔
$tipInterval = 5
# 割草机目录
$grasscutter = 'D:\grasscutter\'
# 原神目录
$genshin = 'D:\Genshin Impact\Genshin Impact Game\YuanShen.exe'
# 端口
$port = '54321'

function setProxyRegV {
    param ($stat)
    Set-ItemProperty 'HKCU:\Software\Microsoft\Windows\CurrentVersion\Internet Settings' -Name ProxyEnable -Value $stat
}
function disableProxy {
    setProxyRegV 0
    Stop-Process -Force -Confirm:$false -Name mitmdump
}
Set-Location $grasscutter
Start-Process -FilePath $genshin
if (-not $?) { disableProxy; Exit }
Start-Process java -WindowStyle Normal -ArgumentList '-jar grasscutter.jar'
# repalce 'Normal' with 'Minimized' to gain a Minimized window
Start-Process -WindowStyle Hidden mitmdump -ArgumentList "-s proxy.py --ssl-insecure --set block_global=false --listen-port $port"
setProxyRegV 1
for ($t = $loginTimeout; $t -gt 0; $t -= $tipInterval) {
    Clear-Host
    Write-Host "Please login in $t seconds."
    Start-Sleep -Seconds $tipInterval
}
disableProxy
```

### 我自己用的代码
- 自我隐藏（最小化）、隐藏mitmdump窗口
- 登录超时80秒，并删去了倒计时代码
- 能简则简
- 服务端控制台不隐藏，代码刷屏很舒适

```POWERSHELL
param($self = $true)
if ($self) { Start-Process PowerShell -ArgumentList ('-WindowStyle Minimized "{0}" $false' -f $MyInvocation.MyCommand.Definition); Exit }
Set-Location 'D:\grasscutter\'
function setProxyRegV {
    param($stat)
    Set-ItemProperty 'HKCU:\Software\Microsoft\Windows\CurrentVersion\Internet Settings' -Name ProxyEnable -Value $stat
}
function disableProxy {
    setProxyRegV 0
    Stop-Process -Force -Confirm:$false -Name mitmdump
}
Start-Process -FilePath 'D:\Genshin Impact\Genshin Impact Game\YuanShen.exe'
if (-not $?) { disableProxy; Exit }
Start-Process java -ArgumentList '-jar grasscutter.jar'
Start-Process -WindowStyle Hidden mitmdump -ArgumentList '-s proxy.py --ssl-insecure --set block_global=false --listen-port 54321'
setProxyRegV 1
Start-Sleep -Seconds 80
disableProxy
```

**Comment by loktindyi** [comment]
原来有个`start.cmd`，大概看了一下，至少大概是没有隐藏窗口的功能的
（为什么能写那么一大堆啊）
有点看不明白了，似乎也没有自动关闭代理，可能，看不下去了
总之我写的这玩意应该算是有点用

**Reply by zhaodice** [reply]
我也写了个类似的，不过用的是易语言XD

**Reply by loktindyi** [reply]
> 我也写了个类似的，不过用的是易语言XD

我不太了解，易语言现在还报毒吗=P

**Reply by zhaodice** [reply]
> > 我也写了个类似的，不过用的是易语言XD
> 
> 我不太了解，易语言现在还报毒吗=P

当然报毒，但不影响使用XD
